cpp/daal/src/algorithms/kmeans/kmeans_lloyd_batch_impl.i

```diff
@@ -241,30 +213,61 @@ Status KMeansBatchKernel<method, algorithmFPType, cpu>::compute(const NumericTab
 
     if (par->resultsToEvaluate & computeAssignments || par->assignFlag || par->resultsToEvaluate & computeExactObjectiveFunction)
     {
-        PostProcessing<method, algorithmFPType, cpu>::computeAssignments(p, nClusters, clusters, ntData, catCoef.get(), assignmetsNT, blockSize);
+        PostProcessing<method, algorithmFPType, cpu>::computeAssignments(p, nClusters, clusters, ntData, nullptr, assignmetsNT, blockSize);
     }
 
-    WriteOnlyRows<algorithmFPType, cpu> mtTarget(*const_cast<NumericTable *>(r[2]), 0, 1);
-    DAAL_CHECK_BLOCK_STATUS(mtTarget);
     if (par->resultsToEvaluate & computeExactObjectiveFunction)
     {
+        WriteOnlyRows<algorithmFPType, cpu> mtTarget(*const_cast<NumericTable *>(r[2]), 0, 1);
+        DAAL_CHECK_BLOCK_STATUS(mtTarget);
         algorithmFPType exactTargetFunc = algorithmFPType(0);
-        PostProcessing<method, algorithmFPType, cpu>::computeExactObjectiveFunction(p, nClusters, clusters, ntData, catCoef.get(), assignmetsNT,
+        PostProcessing<method, algorithmFPType, cpu>::computeExactObjectiveFunction(p, nClusters, clusters, ntData, nullptr, assignmetsNT,
                                                                                     exactTargetFunc, blockSize);
 
         *mtTarget.get() = exactTargetFunc;
     }
-    else
-    {
-        *mtTarget.get() = oldTargetFunc;
-    }
 
     WriteOnlyRows<int, cpu> mtIterations(*const_cast<NumericTable *>(r[3]), 0, 1);
     DAAL_CHECK_BLOCK_STATUS(mtIterations);
     *mtIterations.get() = kIter;
     return (!result) ? s : services::Status(services::ErrorMemoryCopyFailedInternal);
 }
 
+template <Method method, typename algorithmFPType, CpuType cpu>
+Status KMeansBatchKernel<method, algorithmFPType, cpu>::infer(const NumericTable * const * a, const NumericTable * const * r, const Parameter * par)
+{
+    NumericTable * ntData  = const_cast<NumericTable *>(a[0]);
+    const size_t n         = ntData->getNumberOfRows();
+    const size_t p         = ntData->getNumberOfColumns();
+    const size_t nClusters = par->nClusters;
+
+    NumericTable * assignmetsNT = const_cast<NumericTable *>(r[0]);
+    size_t blockSize            = 0;
+    DAAL_SAFE_CPU_CALL((blockSize = BSHelper<method, algorithmFPType, cpu>::kmeansGetBlockSize(n, p, nClusters)), (blockSize = 512))
+
+    ReadRows<algorithmFPType, cpu> mtInClusters(*const_cast<NumericTable *>(a[1]), 0, nClusters);
+    DAAL_CHECK_BLOCK_STATUS(mtInClusters);
+    algorithmFPType * inClusters = const_cast<algorithmFPType *>(mtInClusters.get());
+
+    if (par->resultsToEvaluate & computeAssignments || par->resultsToEvaluate & computeExactObjectiveFunction)
+    {
+        PostProcessing<method, algorithmFPType, cpu>::computeAssignments(p, nClusters, inClusters, ntData, nullptr, assignmetsNT, blockSize);
+    }
+
+    if (par->resultsToEvaluate & computeExactObjectiveFunction)
+    {
+        WriteOnlyRows<algorithmFPType, cpu> mtTarget(*const_cast<NumericTable *>(r[1]), 0, 1);
```

Comment:
What is r? And why is r[1] here used when r[2] is used above?

Reply:
r is pointer to result arrays. You can find what is it in infer_cpp file